File: crates/connect-mirror/src/sync_revalidator.rs
use super::*;

impl DirectoryMirror {
    pub(super) fn revalidate(
        &self,
        plan: &MirrorSyncPlan,
        state: Option<&DurableMirrorState>,
    ) -> Result<(), MirrorError> {
        if state.map(|value| value.generation).unwrap_or(0) != plan.checkpoint_generation
            || state.map(|value| value.cursor) != plan.base_cursor
        {
            return Err(MirrorError::new(
                "sync_plan_stale",
                "Durable checkpoint changed after inspection.",
            ));
        }
        for action in &plan.actions {
            if !action.depends_on().is_empty() {
                continue;
            }
            match action {
                SyncAction::WriteLocal {
                    target,
                    expected_local,
                    expected_path_owner,
                    ..
                }
                | SyncAction::DeleteLocal {
                    target,
                    expected_local,
                    expected_path_owner,
                    ..
                } => {
                    self.revalidate_expected(expected_local)?;
                    self.revalidate_at(&target.path, expected_path_owner)?;
                }
                SyncAction::MoveLocal {
                    source,
                    target_path,
                    expected_source_owner,
                    expected_target_owner,
                    ..
                } => {
                    self.revalidate_at(&source.path, expected_source_owner)?;
                    self.revalidate_at(target_path, expected_target_owner)?;
                }
                SyncAction::PutRemote { expected_local, .. }
                | SyncAction::MoveRemote { expected_local, .. }
                | SyncAction::RecordConflict {
                    local: expected_local,
                    ..
                } => self.revalidate_expected(expected_local)?,
                SyncAction::DeleteRemote {
                    target,
                    expected_local,
                    ..
                } => self.revalidate_at(&target.path, expected_local)?,
                SyncAction::AdvanceCheckpoint { .. } => {}
            }
        }
        Ok(())
    }

    pub(super) fn revalidate_at(
        &self,
        path: &str,
        expected: &ExpectedObjectState,
    ) -> Result<(), MirrorError> {
        match expected {
            ExpectedObjectState::Absent => {
                if self.read_file(path)?.is_some() || self.file_digest(path)?.is_some() {
                    return Err(MirrorError::new(
                        "sync_plan_stale",
                        format!("{path} is no longer vacant."),
                    ));
                }
            }
            ExpectedObjectState::Exact { object } if object.path == path => {
                self.revalidate_expected(expected)?
            }
            _ => {
                return Err(MirrorError::new(
                    "sync_plan_stale",
                    format!("{path} has a different owner."),
                ))
            }
        }
        Ok(())
    }

    pub(super) fn revalidate_expected(
        &self,
        expected: &ExpectedObjectState,
    ) -> Result<(), MirrorError> {
        let ExpectedObjectState::Exact { object } = expected else {
            return Ok(());
        };
        let matches = match object.entity {
            SyncObjectKind::File => {
                self.file_digest(&object.path)?.as_deref() == Some(&object.payload_revision)
            }
            _ => self.read_file(&object.path)?.is_some_and(|value| {
                format!("sha256:{}", digest(&value)) == object.payload_revision
            }),
        };
        if !matches {
            return Err(MirrorError::new(
                "sync_plan_stale",
                format!("{} no longer matches inspected bytes.", object.path),
            ));
        }
        Ok(())
    }

    pub(super) fn matches_ref(&self, object: &SyncObjectRef) -> Result<bool, MirrorError> {
        match object.entity {
            SyncObjectKind::File => Ok(self.file_digest(&object.path)?.as_deref()
                == Some(object.payload_revision.as_str())),
            _ => Ok(self.read_file(&object.path)?.is_some_and(|value| {
                format!("sha256:{}", digest(&value)) == object.payload_revision
            })),
        }
    }
}

File: packages/sync/src/sync-revalidator.ts
import { SyncError } from "./sync-error.js";
import { runtimeDocumentRevision } from "./mirror-format.js";
import type { MirrorFileSystem, MirrorRuntime, MirrorState } from "./mirror-state.js";
import type { ExpectedObjectState, SyncAction, SyncObjectRef } from "./sync-model.js";
import type { ReconciliationPlan } from "./sync-planner.js";

/**
 * Last read-only boundary before preparation. A prepared batch thereafter owns
 * immutable payloads and is recovered by replay, never by consulting live paths.
 */
export class PlanRevalidator {
  constructor(
    private readonly fileSystem: MirrorFileSystem,
    private readonly runtime: MirrorRuntime
  ) {}

  async validate(plan: ReconciliationPlan, state: MirrorState | null): Promise<void> {
    if (
      (state?.generation ?? 0) !== plan.checkpoint_generation
      || (state?.cursor ?? null) !== plan.base_cursor
      || (state?.scope_epoch ?? plan.scope_epoch) !== plan.scope_epoch
    ) throw stale("The durable checkpoint changed after inspection.");
    for (const action of plan.actions) {
      if (action.depends_on.length === 0) await this.validateAction(action);
    }
  }

  private async validateAction(action: SyncAction): Promise<void> {
    switch (action.command) {
      case "write_local":
        await this.validateExpected(action.expected_local);
        await this.validateExpectedAt(action.target.path, action.expected_path_owner);
        return;
      case "delete_local":
        await this.validateExpected(action.expected_local);
        await this.validateExpectedAt(action.target.path, action.expected_path_owner);
        return;
      case "move_local":
        await this.validateExpectedAt(action.source.path, action.expected_source_owner);
        await this.validateExpectedAt(action.target_path, action.expected_target_owner);
        return;
      case "put_remote":
      case "move_remote":
        await this.validateExpected(action.expected_local);
        return;
      case "delete_remote":
        await this.validateExpectedAt(action.target.path, action.expected_local);
        return;
      case "record_conflict":
        await this.validateExpected(action.local);
        return;
      case "clear_conflict":
        await this.validateExpected(action.expected_local);
        return;
      case "advance_checkpoint":
        return;
    }
  }

  async validateExpectedAt(path: string, expected: ExpectedObjectState): Promise<void> {
    if (expected.state === "absent") {
      if (await this.fileSystem.exists(path)) throw stale(`${path} is no longer vacant.`);
      return;
    }
    if (
      expected.object.path !== path
      || !await matchesLocalRef(this.fileSystem, this.runtime, expected.object, true)
    ) {
      throw stale(`${path} no longer has the inspected owner and bytes.`);
    }
  }

  async validateExpected(expected: ExpectedObjectState): Promise<void> {
    if (expected.state === "absent") return;
    if (!await matchesLocalRef(this.fileSystem, this.runtime, expected.object, true)) {
      throw stale(`${expected.object.path} no longer matches the inspected bytes.`);
    }
  }

}

export async function matchesLocalRef(
  fileSystem: MirrorFileSystem,
  runtime: MirrorRuntime,
  ref: SyncObjectRef,
  revalidating: boolean
): Promise<boolean> {
  if (ref.entity === "file") {
    const info = await fileSystem.inspectBinary(ref.path);
    return !!info
      && info.content_digest === ref.payload_revision
      && (ref.size === undefined || info.size === ref.size);
  }
  let read;
  try {
    read = await fileSystem.readText(ref.path);
  } catch (error) {
    if (
      error instanceof SyncError
      && error.code === "file_read_failed"
      && revalidating
    ) throw stale(error.message);
    throw error;
  }
  return (typeof read === "string" ? runtimeDocumentRevision(read, runtime) : read?.revision)
    === ref.payload_revision;
}

export function stale(message: string): SyncError {
  return new SyncError("sync_plan_stale", message);
}
